core/src/main/java/io/grpc/internal/ConnectivityStateManager.java

```diff
@@ -16,66 +16,103 @@
+import static com.google.common.base.Preconditions.checkState;
+
 import io.grpc.ConnectivityState;
+import io.grpc.ManagedChannel;
 import java.util.ArrayList;
 import java.util.concurrent.Executor;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
 import javax.annotation.concurrent.NotThreadSafe;
 
 /**
- * Book-keeps the connectivity state and callbacks.
+ * Manages connectivity states of the channel. Used for {@link ManagedChannel#getState} to read the
+ * current state of the channel, for {@link ManagedChannel#notifyWhenStateChanged} to add
+ * listeners to state change events, and for {@link io.grpc.LoadBalancer.Helper#updateBalancingState
+ * LoadBalancer.Helper#updateBalancingState} to update the state and run the {@link #gotoState}s.
  */
 @NotThreadSafe
-class ConnectivityStateManager {
-  private ArrayList<StateCallbackEntry> callbacks;
-
-  private ConnectivityState state;
+final class ConnectivityStateManager {
+  private ArrayList<Listener> listeners = new ArrayList<Listener>();
 
-  ConnectivityStateManager(ConnectivityState initialState) {
-    state = initialState;
-  }
+  private volatile ConnectivityState state = ConnectivityState.IDLE;
 
+  /**
+   * Adds a listener for state change event.
+   *
+   * <p>The {@code executor} must be one that can run RPC call listeners.
+   */
   void notifyWhenStateChanged(Runnable callback, Executor executor, ConnectivityState source) {
-    StateCallbackEntry callbackEntry = new StateCallbackEntry(callback, executor);
+    checkNotNull(callback, "callback");
+    checkNotNull(executor, "executor");
+    checkNotNull(source, "source");
+
+    Listener stateChangeListener = new Listener(callback, executor);
     if (state != source) {
-      callbackEntry.runInExecutor();
+      stateChangeListener.runInExecutor();
     } else {
-      if (callbacks == null) {
-        callbacks = new ArrayList<StateCallbackEntry>();
-      }
-      callbacks.add(callbackEntry);
+      listeners.add(stateChangeListener);
     }
   }
 
-  // TODO(zhangkun83): return a runnable in order to escape transport set lock, in case direct
-  // executor is used?
-  void gotoState(ConnectivityState newState) {
+  /**
+   * Connectivity state is changed to the specified value. Will trigger some notifications that have
+   * been registered earlier by {@link ManagedChannel#notifyWhenStateChanged}.
+   */
+  void gotoState(@Nonnull ConnectivityState newState) {
+    checkNotNull(newState, "newState");
+    gotoNullableState(newState);
+  }
+
+  private void gotoNullableState(@Nullable ConnectivityState newState) {
+    checkState(state != null, "ConnectivityStateManager is already disabled");
     if (state != newState) {
-      if (state == ConnectivityState.SHUTDOWN) {
-        throw new IllegalStateException("Cannot transition out of SHUTDOWN to " + newState);
-      }
+      checkState(
+          state != ConnectivityState.SHUTDOWN,
+          "Cannot transition out of SHUTDOWN to " + newState);
```

Comment:
What if the channel is shutdown immediately after created? `LoadBalancer` would not have been created, and channel would stay in `IDLE` forever. This is not right.

Since channel outlives the balancer, `SHUTDOWN` state should be triggered by channel, not the balancer. An interesting case is, if balancer reports `SHUTDOWN` but channel is not shutdown yet, what should we do?